#! /bin/sh /usr/share/dpatch/dpatch-run
## 10no_one_line_visual.dpatch by  <hesso@pool.math.tu-berlin.de>
##
## DP: Catch segfaults when the screen is only one line high.

@DPATCH@
diff -Naur nvi-1.81.6.orig/vi/vi.c nvi-1.81.6/vi/vi.c
--- nvi-1.81.6.orig/vi/vi.c	2007-11-18 17:41:42.000000000 +0100
+++ nvi-1.81.6/vi/vi.c	2008-05-01 18:15:14.000000000 +0200
@@ -974,6 +974,14 @@
 	sp->rows = vip->srows = O_VAL(sp, O_LINES);
 	sp->cols = O_VAL(sp, O_COLUMNS);
 	sp->t_rows = sp->t_minrows = O_VAL(sp, O_WINDOW);
+	/*
+	 * To avoid segfaults on terminals with only one line,
+	 * catch this corner case now and die explicitly.
+	 */
+	if (sp->t_rows == 0) {
+		(void)fprintf(stderr, "Error: Screen too small for visual mode.\n");
+		return 1;
+	}
 	if (sp->rows != 1) {
 		if (sp->t_rows > sp->rows - 1) {
 			sp->t_minrows = sp->t_rows = sp->rows - 1;
